icechunk/src/store.rs

```diff
@@ -428,31 +423,90 @@ impl Store {
-        Ok(futures::stream::iter(parents.into_iter().map(Ok)))
+                res
+            }
+            Ok(NodeSnapshot { node_data: NodeData::Group, .. }) => {
+                // if the prefix is the path to a group we need to discover any nodes with the prefix as node path
+                // listing chunks is unnecessary
+                self.list_metadata_prefix(prefix, true)
+                    .await?
+                    .try_filter_map(|x| async move {
+                        let x = x.trim_end_matches("/zarr.json").to_string();
+                        let res = if x == "zarr.json" {
+                            Some(ListDirItem::Key("zarr.json".to_string()))
+                        } else if x.matches("/").count() == 0 {
+                            Some(ListDirItem::Prefix(x))
+                        } else {
+                            None
+                        };
+                        Ok(res)
+                    })
+                    .try_collect::<Vec<_>>()
+                    .await?
+            }
+            Err(_) => {
+                // Finally if there is no node at the prefix we look for chunks belonging to an ancestor node
+                // FIXME: This iterates over every chunk of the array and is wasteful
+                let node = session.get_closest_ancestor_node(&path).await;
+                if let Ok(node) = node {
+                    let node_path = node.path.clone();
+                    session
+                        .array_chunk_iterator(&node.path)
+                        .await
+                        .try_filter_map(|chunk| async {
+                            let chunk_key = Key::Chunk {
+                                node_path: node_path.clone(),
+                                coords: chunk.coord,
+                            }
+                            .to_string();
+                            let res = if is_prefix_match(&chunk_key, prefix) {
+                                Some({
+                                    let trimmed = chunk_key
+                                        .trim_start_matches(prefix)
+                                        .trim_start_matches("/");
+                                    if let Some((chunk_prefix, _)) =
+                                        trimmed.split_once("/")
+                                    {
+                                        ListDirItem::Prefix(chunk_prefix.to_string())
+                                    } else {
+                                        ListDirItem::Key(trimmed.to_string())
+                                    }
+                                })
+                            } else {
+                                None
+                            };
+                            Ok(res)
+                        })
+                        .try_collect::<HashSet<_>>()
+                        .await?
+                        .into_iter()
+                        .collect::<Vec<_>>()
```

Comment:
very yuck, I guess I could just create the stream in each arm of the match instead of returning a Vec

Reply:
no that doesn't work becausse HashSet.into_iter() and Vec.into_iter() are different types :/